### PR TITLE
Recognize mod-free helper calls in loop hoisting

### LINKED ISSUE #6901: JIT: Loop hoisting is overly-conservative with helper calls
Loop hoisting has code to avoid illegally re-ordering side-effects, which includes [a check for calls](https://github.com/dotnet/coreclr/blob/03ff9b0b6b7fc7a037f4d03a904ed5da8a208dbb/src/jit/optimizer.cpp#L6229) that is overly-general.  The check looks at the `GTF_CALL` flag, which (A) will be set for the entire expression containing a call as a sub-expression (which is unnecessary because this code is walking the whole tree), and (B) doesn't distinguish non-side-effecting helper calls, like we do in [value-numbering](https://github.com/dotnet/coreclr/blob/03ff9b0b6b7fc7a037f4d03a904ed5da8a208dbb/src/jit/valuenum.cpp#L7216) and in [the code just above that identifies hoistable calls](https://github.com/dotnet/coreclr/blob/03ff9b0b6b7fc7a037f4d03a904ed5da8a208dbb/src/jit/optimizer.cpp#L6162-L6181).  This blocks hoisting of the `m_arr` accesses (because of the `GetStaticBase` helper call that precedes them) in the following code, even with a fix for #6900 in place:

```c#
using System;
using System.Diagnostics;

class ArrayPerf
{

    private static int[] m_arr;
    private const int MAX_ARRAY_SIZE = 4096;

    private static readonly int DIM_1 = MAX_ARRAY_SIZE;

    static void Main(string[] args)
    {
        long iterations = Int64.Parse(args[0]);

        int value;
        m_arr = new int[DIM_1];

        for (int i = 0; i < DIM_1; i++)
            m_arr[i] = i;

        for (int j = 0; j < DIM_1; j++)
            value = m_arr[j];

        //var sw = Stopwatch.StartNew();

        for (long i = 0; i < iterations; i++)
        {
            for (int j = 0; j < DIM_1; j++)
            {
                value = m_arr[j];
                value = m_arr[j];
                value = m_arr[j];
                value = m_arr[j];
                value = m_arr[j];
                value = m_arr[j];
                value = m_arr[j];
                value = m_arr[j];
                value = m_arr[j];
                value = m_arr[j];
            }
        }

        //sw.Stop();
        //Console.WriteLine(sw.ElapsedMilliseconds);
    }
}
```

### PR DESCRIPTION
**Pull request by JosephTremoulet**
Update the code in `optHoistLoopExprsForTree` that identifies exprs with
memory side-effects to recognize helper calls that don't mutate the heap;
this allows hoisting invariant exprs past such helper calls.

Fixes #6901.

### CODE REVIEW

**Comment by JosephTremoulet** [comment]
@dotnet/jit-contrib PTAL.  This enables more aggressive hoisting in 84 mscorlib methods, and ~250 desktop frameworks methods.

**Comment by pgavlin** [comment]
LGTM.

Any examples of the diffs?

**Comment by JosephTremoulet** [comment]
> Any examples of the diffs?

Sure.  Here's one where we hoist the length load of a bounds check out of a loop that has a newarray operation preceding the bounds check; the newarray helper call was blocking the hoisting without this change.

Before:
```asm
       ...

G_M65471_IG03:
       mov      rcx, 0xD1FFAB1E
       call     CORINFO_HELP_NEWSFAST
       mov      r14, rax
       cmp      ebp, dword ptr [rdi+8]
       jae      G_M65471_IG13
       movsxd   rcx, ebp
       shl      rcx, 4
       lea      rcx, bword ptr [rdi+rcx+16]
       ...
       inc      ebp
       cmp      ebp, ebx
       jl       SHORT G_M65471_IG03
```

After:
```asm
       ...
       mov      r14d, dword ptr [rbx+8]

G_M65471_IG03:
       mov      rcx, 0xD1FFAB1E
       call     CORINFO_HELP_NEWSFAST
       mov      r15, rax
       cmp      ebp, r14d
       jae      G_M65471_IG13
       movsxd   rcx, ebp
       shl      rcx, 4
       lea      rcx, bword ptr [rbx+rcx+16]
       ...
       inc      ebp
       cmp      ebp, edi
       jl       SHORT G_M65471_IG03
```

File: src/jit/optimizer.cpp [hunk 1]
```diff
@@ -6226,9 +6226,28 @@ bool Compiler::optHoistLoopExprsForTree(
         // be hoisted so that they are evaluated in the same order as they would have been in the loop,
         // and therefore throw exceptions in the same order.  (So we don't use GTF_GLOBALLY_VISIBLE_SIDE_EFFECTS
         // here, since that includes exceptions.)
-        if (tree->gtFlags & GTF_CALL)
+        if (tree->IsCall())
         {
-            *pFirstBlockAndBeforeSideEffect = false;
+            // If it's a call, it must be a helper call, and be not mutate the heap.
```
Comment (erozenfeld) [inline src/jit/optimizer.cpp]: not mutate --> not mutating

**Comment by erozenfeld** [comment]
LGTM

File: src/jit/optimizer.cpp (hunk 1 above)
Comment (briansull) [inline src/jit/optimizer.cpp]: and also not mute the heap.

**Comment by briansull** [comment]
LGTM

**Pull request merged**